include/lsst/afw/typehandling/HeteroMap.h

```diff
@@ -0,0 +1,700 @@
+ * values. It does not support unique pointers to Storable because such pointers are read destructively. For
+ * safety reasons, it may not contain references, C-style pointers, or arrays to any type. Due to
+ * implementation restrictions, `const` types (particularly pointers to `const` Storable) are not
+ * currently supported.
+ */
+// TODO: const keys should be possible in C++17 with std::variant
+template <typename K>
+class HeteroMap {
+public:
+    using key_type = K;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    virtual ~HeteroMap() = default;
+
+    /**
+     * Return a reference to the mapped value of the element with key equal to `key`.
+     *
+     * @tparam T the type of the element mapped to `key`
+     * @param key the key of the element to find
+     *
+     * @return a reference to the `T` mapped to `key`, if one exists
+     *
+     * @throws pex::exceptions::OutOfRangeError Thrown if the map does not
+     *         have a `T` with the specified key
+     * @exceptsafe Provides strong exception safety.
+     *
+     * @note This implementation calls @ref unsafeLookup once, then uses templates
+     *       and RTTI to determine if the value is of the expected type.
+     *
+     * @{
+     */
+    template <typename T, typename std::enable_if_t<!IS_SMART_PTR<T>, int> = 0>
+    T& at(Key<K, T> const& key) {
+        // Both casts are safe; see Effective C++, Item 3
+        return const_cast<T&>(static_cast<const HeteroMap&>(*this).at(key));
+    }
+
+    // Can't partially specialize method templates, rely on enable_if to avoid duplicates
+    template <typename T,
+              typename std::enable_if_t<
+                      std::is_fundamental<T>::value || std::is_base_of<std::string, T>::value, int> = 0>
+    T const& at(Key<K, T> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            return boost::get<T const&>(foo);
+        } catch (boost::bad_get const&) {
+            std::stringstream message;
+            message << "Key not found: " << key;
+            throw LSST_EXCEPT(pex::exceptions::OutOfRangeError, message.str());
+        }
+    }
+
+    template <typename T, typename std::enable_if_t<std::is_base_of<Storable, T>::value, int> = 0>
+    T const& at(Key<K, T> const& key) const {
+        static_assert(!std::is_const<T>::value,
+                      "Due to implementation constraints, const keys are not supported.");
+        try {
+            auto foo = unsafeLookup(key.getId());
+            // Don't use pointer-based get, because it won't work after migrating to std::variant
+            // return unique_ptr by reference, so the pointer in internal storage won't get cleared
+            auto& holder = boost::get<std::unique_ptr<Storable> const&>(foo);
```

Comment:
I think we need a strong distinction between `Storable` and `shared_ptr<Storable>` on the C++ side, for the same reason we want an (almost) type-safe map in the first place -- they are different types, and you can get surprising behavior (segfaults!) if you get them mixed up. This is not a problem in Python because I treat any insertion from Python as `shared_ptr<Storable>`, and Python queries don't use typed keys.

If we're keeping `Storable`, then we need reference access to avoid slicing.

Reply:
Over the weekend I realized that you can factor all the `unique_ptr` handling into a separate class, greatly simplifying both the C++ code and the pybind11 wrapper. I've committed those changes after the ones you reviewed.